pkg/storage/ingest/reader.go

```diff
@@ -271,6 +294,9 @@ func (r *PartitionReader) run(ctx context.Context) error {
 // switchToOngoingFetcher switches to the configured ongoing fetcher. This function could be
 // called multiple times.
 func (r *PartitionReader) switchToOngoingFetcher(ctx context.Context) {
+	r.fetcherMtx.Lock()
+	defer r.fetcherMtx.Unlock()
```

Comment:
Here we may the lock for some long period of time, which would block metrics scraping. I've also seen the lock is not taken in `processNextFetches()` and `stopDependencies()` which makes it harder to reason about whether it's required or not. Since `r.client` is not set in the "constructure" but in `start()` to guarantee avoiding any race condition you would actually need a lock around it too.

We need to simplify this locking. I think that we just want to protect is from race conditions on the fetcher and client pointer. We don't need a locking to protect race conditions in the logic, just the pointer. So I think we could add `setClient()`, `getClient()`, `setFetcher()`, `getFetcher()`: these 4 functions just manipulate the fetcher/client pointer wrapped with a lock. Then each time we need to set/get the pointers, we call these functions. No other locking required in the callers.

Reply:
Thanks for the suggestion - I had the same concern but didn't have any better ideas that would not involve a moving too much code around. Your suggestion of the four functions makes sense.

Reply:
Done in https://github.com/grafana/mimir/pull/9855/commits/e14bba531244269d51dde0c0fc0fc47776219516